### DISCUSSION TITLE
terminal base64 encoding result appear to be different from openzeppeling base64 result.

### DISCUSSION THREAD

**Discussion by EngrPips** [Cyfrin/foundry-full-course-cu#266] [General]
So when i ran the test for deployMoodNftTest i was getting test failed and the reason was because my expectedURI didn't match with the actualURI i got back from the encoding done in my DeployMoodNft script. that get me really confuse as i don't understand why the terminal encoding result returned something different from the onchain encoding.

below are the details for my test and results;

## The svg i am encoding 
```
<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" width="500" height="500"><text x="0" y="15" fill="black">Hi! Your browser decoded this</text></svg>
```
## result returned on running the ```base64 -i ./img/example.svg```

```
PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHhtbG5zOnhsaW5rPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5L3hsaW5rIiB3aWR0aD0iNTAwIiBoZWlnaHQ9IjUwMCI+PHRleHQgeD0iMCIgeT0iMTUiIGZpbGw9ImJsYWNrIj5IaSEgWW91ciBicm93c2VyIGRlY29kZWQgdGhpczwvdGV4dD48L3N2Zz4KICAgIAo=
```

## My DeployMoodNft script

```solidity
//SPDX-License-Identifier: MIT
pragma solidity ^ 0.8.18;

import {Script} from "forge-std/Script.sol";
import {MoodNft} from "../src/MoodNft.sol";
import {Base64} from "@openzeppelin/contracts/utils/Base64.sol";

contract DeployMoodNft is Script{
    function run() external returns (MoodNft) {

    }

    function svgToURI(string memory _svg) public pure returns(string memory){
        //note -> we would be passing an actual svg text to this function and it would encode it into base64 for us.
        string memory baseURI = "data:image/svg+xml;base64,";
        string memory svgBase64Encoded = Base64.encode(bytes(string(abi.encodePacked(_svg))));
        return string(abi.encodePacked(baseURI, svgBase64Encoded));
    } 
}
```

## My DeployMoodNftTest Test

```solidity
//SPDX-License-Identifier: MIT

pragma solidity ^ 0.8.18;

import {Test, console} from "forge-std/Test.sol";
import {DeployMoodNft} from "../../script/DeployMoodNft.s.sol";

contract DeployMoodNftTest is Test {
    DeployMoodNft public deployer;
    function setUp() public {
        deployer = new DeployMoodNft();
    }

    function testConvertSvgToURI() public view {
        string memory expectedURI = "data:image/svg+xml;base64,data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHhtbG5zOnhsaW5rPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5L3hsaW5rIiB3aWR0aD0iNTAwIiBoZWlnaHQ9IjUwMCI+PHRleHQgeD0iMCIgeT0iMTUiIGZpbGw9ImJsYWNrIj5IaSEgWW91ciBicm93c2VyIGRlY29kZWQgdGhpczwvdGV4dD48L3N2Zz4KICAgIAo=";

        string memory svg = '<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" width="500" height="500"><text x="0" y="15" fill="black">Hi! Your browser decoded this</text></svg>';

        string memory actualURI = deployer.svgToURI(svg);
        assert(keccak256(abi.encodePacked(actualURI)) == keccak256(abi.encodePacked(expectedURI)));
    }
}
```

## Result returned upon running ```forge test --mt testConvertSvgToURI```

```
[FAIL. Reason: Assertion violated] testConvertSvgToURI() (gas: 26718)
Traces:
  [344019] DeployMoodNftTest::setUp() 
    ├─ [289350] → new DeployMoodNft@0x5615dEB798BB3E4dFa0139dFa1b3D433Cc23b72f
    │   └─ ← 1224 bytes of code
    └─ ← ()

  [26718] DeployMoodNftTest::testConvertSvgToURI() 
    ├─ [17746] DeployMoodNft::svgToURI(<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" width="500" height="500"><text x="0" y="15" fill="black">Hi! Your browser decoded this</text></svg>) [staticcall]
    │   └─ ← data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHhtbG5zOnhsaW5rPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5L3hsaW5rIiB3aWR0aD0iNTAwIiBoZWlnaHQ9IjUwMCI+PHRleHQgeD0iMCIgeT0iMTUiIGZpbGw9ImJsYWNrIj5IaSEgWW91ciBicm93c2VyIGRlY29kZWQgdGhpczwvdGV4dD48L3N2Zz4=
    └─ ← "Assertion violated"

Test result: FAILED. 0 passed; 1 failed; 0 skipped; finished in 8.16ms
Ran 1 test suites: 0 tests passed, 1 failed, 0 skipped (1 total tests)

Failing tests:
Encountered 1 failing test in test/unit/DeployMoodNftTest.t.sol:DeployMoodNftTest
[FAIL. Reason: Assertion violated] testConvertSvgToURI() (gas: 26718)

Encountered a total of 1 failing tests, 0 tests succeeded
```
## The above shows that the actualURI string value is the value below
```PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHhtbG5zOnhsaW5rPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5L3hsaW5rIiB3aWR0aD0iNTAwIiBoZWlnaHQ9IjUwMCI+PHRleHQgeD0iMCIgeT0iMTUiIGZpbGw9ImJsYWNrIj5IaSEgWW91ciBicm93c2VyIGRlY29kZWQgdGhpczwvdGV4dD48L3N2Zz4=```

## from all of the above, it was clear why the test failed to me because the tokenURI i got back from doing terminal encoding is different from the tokenURI value that my DeployMoodNft script returned but what confuses me is why are the result different when in fact i am encoding exactly the same value.

```
Terminal encoding result (PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHhtbG5zOnhsaW5rPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5L3hsaW5rIiB3aWR0aD0iNTAwIiBoZWlnaHQ9IjUwMCI+PHRleHQgeD0iMCIgeT0iMTUiIGZpbGw9ImJsYWNrIj5IaSEgWW91ciBicm93c2VyIGRlY29kZWQgdGhpczwvdGV4dD48L3N2Zz4KICAgIAo=)  != 
DeployMoodNft script encoding result
(PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHhtbG5zOnhsaW5rPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5L3hsaW5rIiB3aWR0aD0iNTAwIiBoZWlnaHQ9IjUwMCI+PHRleHQgeD0iMCIgeT0iMTUiIGZpbGw9ImJsYWNrIj5IaSEgWW91ciBicm93c2VyIGRlY29kZWQgdGhpczwvdGV4dD48L3N2Zz4=)
```
i would appreciate any help at all @PatrickAlphaC

**Comment by shikhar229169** [comment]
I have already answer this in previous same issue: https://github.com/Cyfrin/foundry-full-course-f23/discussions/248

**Reply by EngrPips** [reply]
oh really, i tried looking up if someone have asked previously but couldn't find any, guess i didn't search the correct keyword. i would check that now

**Reply by EngrPips** [reply]
but honestly i am still curious to why @PatrickAlphaC  test pass and ours fails, he encoded in command line and still got same value returned from the onchain encoding.
@shikhar229169  do you have anything to say to me concerning that.

**Reply by shikhar229169** [reply]
In the example.svg file, everything should be in same line, and there should be no new line after the first line.
Also ensure that there is no single space in the end in that svg file.
Maybe this is creating problem.

**Reply by shikhar229169** [reply]
When I run ```base64 -i ./img/example.svg``` in the terminal for the example.svg it gave me 
```PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHhtbG5zOnhsaW5rPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5L3hsaW5rIiB3aWR0aD0iNTAwIiBoZWlnaHQ9IjUwMCI+PHRleHQgeD0iMCIgeT0iMTUiIGZpbGw9ImJsYWNrIj5IaSEgWW91ciBicm93c2VyIGRlY29kZWQgdGhpczwvdGV4dD48L3N2Zz4=```

**Reply by PatrickAlphaC** [reply]
So spaces, new lines, etc all will make your base64 encoded file slightly different. It might be easier to just verify that what OZ is giving you is right, and then use that in your tests

**Reply by AztecWarlord** [reply]
I also was stuck until I reviewed issue: https://github.com/Cyfrin/foundry-full-course-f23/discussions/248 then I converted the ```example.svg``` file to a single line then ran the ```Base64 -i``` command and copy pasted result to ```DeployMoodNftTest.t.sol``` and the test passed. I know issue:248 explains this but, I feel this answer will be simplified for beginners.

**Reply by AztecWarlord** [reply]
just saw this lol.

**Reply by 0xTagger** [reply]
how do I convert it to 1 single line? I've tried merge lines, join lines, and toggle word wrap and no matter what I do it goes to 2 lines once I save the updated file. 

<img width="938" alt="image" src="https://github.com/Cyfrin/foundry-full-course-f23/assets/114359465/bf98ef34-47d4-41d9-9e4e-0fccc8263e4a">

**Comment by 0xSimeon** [comment]
Same issue. Used (https://www.base64encode.org/) to encode then pasted it as `expectedUri` and it worked.

**Reply by SCPassion** [reply]
Wow. Thanks for your link. This works surprisingly well.

**Comment by Lourd03** [comment]
I'm new in the programming space how would you advise is go about this courses??

I've been trying but keep hitting stone edge.

**Reply by 0xSimeon** [reply]
I recommend taking CS50 to give you foundation in computer programming here: https://www.edx.org/cs50
